[JIT] Create torch._C._jit submodule for JIT bindings

SplitInfinity opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#52158 [JIT] Create torch._C._jit submodule for JIT bindings**

**Summary**
Everything inside `torch._C` is currently added to `torch` (with some
policies to exclude names beginning with `_`, etc.). As a result, some
of the JIT Python IR bindings are inadvertently exposed as `torch.XX`.
This commit adds a `torch._C._jit` submodule for all JIT bindings to
preclude JIT bindings being added to `torch`.

**Test Plan**
Continuous integration.

**Fixes**
This commit fixes #51691.

Differential Revision: [D26628027](https://our.internmc.facebook.com/intern/diff/D26628027)

Commit:
[JIT] Create torch._C._jit submodule for JIT bindings

**Summary**
Everything inside `torch._C` is currently added to `torch` (with some
policies to exclude names beginning with `_`, etc.). As a result, some
of the JIT Python IR bindings are inadvertently exposed as `torch.XX`.
This commit adds a `torch._C._jit` submodule for all JIT bindings to
preclude JIT bindings being added to `torch`.

**Test Plan**
Continuous integration.

**Fixes**
This commit fixes #51691.

[ghstack-poisoned]

Commit:
Update on "[JIT] Create torch._C._jit submodule for JIT bindings"


**Summary**
Everything inside `torch._C` is currently added to `torch` (with some
policies to exclude names beginning with `_`, etc.). As a result, some
of the JIT Python IR bindings are inadvertently exposed as `torch.XX`.
This commit adds a `torch._C._jit` submodule for all JIT bindings to
preclude JIT bindings being added to `torch`.

**Test Plan**
Continuous integration.

**Fixes**
This commit fixes #51691.

[ghstack-poisoned]

eellison reviewed (commented):
This is a good end state, that we should have done from the start. However I do want to point out this will break a huge number of APIs that users have relied on (perhaps they shouldnt have).


https://github.com/apache/tvm/blob/main/python/tvm/relay/frontend/pytorch.py#L2699, https://github.com/apple/coremltools/blob/6245e289187c95d32b6ff65bc6348310ebbdf71a/coremltools/converters/mil/frontend/torch/converter.py#L267, and at times we've even recommended they use these APIs: https://github.com/pytorch/pytorch/issues/38342#issuecomment-644324034.


I think the jit passes, and the jit execution (profiler vs legacy) APIs should probably be duplicated in the torch._C namespace to minimize churn

SplitInfinity commented:
Yeah, I pointed that out in the issue but I like this change. I can update the PR to include the bindings (appropriately named) in both `torch._C` and `torch._C._jit`. But would we eventually get rid of the `torch._C` JIT bindings? Or keep both?

eellison commented:
Sorry, I miscommented. I meant just the _jit_set_profiler* apis, and the `_jit_pass_*` apis should be duplicated

Commit:
Update on "[JIT] Create torch._C._jit submodule for JIT bindings"


**Summary**
Everything inside `torch._C` is currently added to `torch` (with some
policies to exclude names beginning with `_`, etc.). As a result, some
of the JIT Python IR bindings are inadvertently exposed as `torch.XX`.
This commit adds a `torch._C._jit` submodule for all JIT bindings to
preclude JIT bindings being added to `torch`.

**Test Plan**
Continuous integration.

**Fixes**
This commit fixes #51691.

[ghstack-poisoned]

Commit:
Update on "[JIT] Create torch._C._jit submodule for JIT bindings"


**Summary**
Everything inside `torch._C` is currently added to `torch` (with some
policies to exclude names beginning with `_`, etc.). As a result, some
of the JIT Python IR bindings are inadvertently exposed as `torch.XX`.
This commit adds a `torch._C._jit` submodule for all JIT bindings to
preclude JIT bindings being added to `torch`.

**Test Plan**
Continuous integration.

**Fixes**
This commit fixes #51691.

[ghstack-poisoned]

Commit:
Update on "[JIT] Create torch._C._jit submodule for JIT bindings"


**Summary**
Everything inside `torch._C` is currently added to `torch` (with some
policies to exclude names beginning with `_`, etc.). As a result, some
of the JIT Python IR bindings are inadvertently exposed as `torch.XX`.
This commit adds a `torch._C._jit` submodule for all JIT bindings to
preclude JIT bindings being added to `torch`.

**Test Plan**
Continuous integration.

**Fixes**
This commit fixes #51691.

[ghstack-poisoned]

Commit:
Update on "[JIT] Create torch._C._jit submodule for JIT bindings"


**Summary**
Everything inside `torch._C` is currently added to `torch` (with some
policies to exclude names beginning with `_`, etc.). As a result, some
of the JIT Python IR bindings are inadvertently exposed as `torch.XX`.
This commit adds a `torch._C._jit` submodule for all JIT bindings to
preclude JIT bindings being added to `torch`.

**Test Plan**
Continuous integration.

**Fixes**
This commit fixes #51691.

[ghstack-poisoned]

Commit:
Update on "[JIT] Create torch._C._jit submodule for JIT bindings"


**Summary**
Everything inside `torch._C` is currently added to `torch` (with some
policies to exclude names beginning with `_`, etc.). As a result, some
of the JIT Python IR bindings are inadvertently exposed as `torch.XX`.
This commit adds a `torch._C._jit` submodule for all JIT bindings to
preclude JIT bindings being added to `torch`.

**Test Plan**
Continuous integration.

**Fixes**
This commit fixes #51691.

[ghstack-poisoned]

Commit:
Update on "[JIT] Create torch._C._jit submodule for JIT bindings"


**Summary**
Everything inside `torch._C` is currently added to `torch` (with some
policies to exclude names beginning with `_`, etc.). As a result, some
of the JIT Python IR bindings are inadvertently exposed as `torch.XX`.
This commit adds a `torch._C._jit` submodule for all JIT bindings to
preclude JIT bindings being added to `torch`.

**Test Plan**
Continuous integration.

**Fixes**
This commit fixes #51691.

[ghstack-poisoned]

Commit:
Update on "[JIT] Create torch._C._jit submodule for JIT bindings"


**Summary**
Everything inside `torch._C` is currently added to `torch` (with some
policies to exclude names beginning with `_`, etc.). As a result, some
of the JIT Python IR bindings are inadvertently exposed as `torch.XX`.
This commit adds a `torch._C._jit` submodule for all JIT bindings to
preclude JIT bindings being added to `torch`.

**Test Plan**
Continuous integration.

**Fixes**
This commit fixes #51691.

[ghstack-poisoned]

Commit:
Update on "[JIT] Create torch._C._jit submodule for JIT bindings"


**Summary**
Everything inside `torch._C` is currently added to `torch` (with some
policies to exclude names beginning with `_`, etc.). As a result, some
of the JIT Python IR bindings are inadvertently exposed as `torch.XX`.
This commit adds a `torch._C._jit` submodule for all JIT bindings to
preclude JIT bindings being added to `torch`.

**Test Plan**
Continuous integration.

**Fixes**
This commit fixes #51691.

[ghstack-poisoned]

Commit:
Update on "[JIT] Create torch._C._jit submodule for JIT bindings"


**Summary**
Everything inside `torch._C` is currently added to `torch` (with some
policies to exclude names beginning with `_`, etc.). As a result, some
of the JIT Python IR bindings are inadvertently exposed as `torch.XX`.
This commit adds a `torch._C._jit` submodule for all JIT bindings to
preclude JIT bindings being added to `torch`.

**Test Plan**
Continuous integration.

**Fixes**
This commit fixes #51691.

[ghstack-poisoned]

Commit:
Update on "[JIT] Create torch._C._jit submodule for JIT bindings"


**Summary**
Everything inside `torch._C` is currently added to `torch` (with some
policies to exclude names beginning with `_`, etc.). As a result, some
of the JIT Python IR bindings are inadvertently exposed as `torch.XX`.
This commit adds a `torch._C._jit` submodule for all JIT bindings to
preclude JIT bindings being added to `torch`.

**Test Plan**
Continuous integration.

**Fixes**
This commit fixes #51691.

[ghstack-poisoned]

Commit:
Update on "[JIT] Create torch._C._jit submodule for JIT bindings"


**Summary**
Everything inside `torch._C` is currently added to `torch` (with some
policies to exclude names beginning with `_`, etc.). As a result, some
of the JIT Python IR bindings are inadvertently exposed as `torch.XX`.
This commit adds a `torch._C._jit` submodule for all JIT bindings to
preclude JIT bindings being added to `torch`.

**Test Plan**
Continuous integration.

**Fixes**
This commit fixes #51691.

[ghstack-poisoned]

Commit:
Update on "[JIT] Create torch._C._jit submodule for JIT bindings"


**Summary**
Everything inside `torch._C` is currently added to `torch` (with some
policies to exclude names beginning with `_`, etc.). As a result, some
of the JIT Python IR bindings are inadvertently exposed as `torch.XX`.
This commit adds a `torch._C._jit` submodule for all JIT bindings to
preclude JIT bindings being added to `torch`.

**Test Plan**
Continuous integration.

**Fixes**
This commit fixes #51691.

[ghstack-poisoned]

Commit:
Update on "[JIT] Create torch._C._jit submodule for JIT bindings"


**Summary**
Everything inside `torch._C` is currently added to `torch` (with some
policies to exclude names beginning with `_`, etc.). As a result, some
of the JIT Python IR bindings are inadvertently exposed as `torch.XX`.
This commit adds a `torch._C._jit` submodule for all JIT bindings to
preclude JIT bindings being added to `torch`.

**Test Plan**
Continuous integration.

**Fixes**
This commit fixes #51691.

[ghstack-poisoned]

Commit:
Update on "[JIT] Create torch._C._jit submodule for JIT bindings"


**Summary**
Everything inside `torch._C` is currently added to `torch` (with some
policies to exclude names beginning with `_`, etc.). As a result, some
of the JIT Python IR bindings are inadvertently exposed as `torch.XX`.
This commit adds a `torch._C._jit` submodule for all JIT bindings to
preclude JIT bindings being added to `torch`.

**Test Plan**
Continuous integration.

**Fixes**
This commit fixes #51691.

[ghstack-poisoned]

SplitInfinity commented:
@albanD @eellison 

Okay, tests are finally passing. I had to adding some bindings to both `torch._C` and `torch._C._jit` and leave some only in `torch._C`, which I will deal with in subsequent PRs.

albanD commented on torch/csrc/jit/python/init.cpp:
```diff
@@ -132,170 +132,499 @@ TORCH_API void runJITCPPTests();
       m, "IODescriptor"); // NOLINT(bugprone-unused-raii)
 
-  m.def("_jit_init", loadPythonClasses)
+  // Register pass and profiler functions on both torch._C and torch.C._jit
```
nit: you can open an issue and link it here for the cleanup of this in case someone reads this code and is interested in doing it.

SplitInfinity replied:
I created #52720 and added a link in the comments. I'm not sure how much can be done here because we would need to find all the external callsites for these passes and fix them...

albanD commented on torch/csrc/jit/python/init.cpp:
```diff
@@ -1308,18 +1358,21 @@ void initJITBindings(PyObject* module) {
+  initPythonCustomClassBindings(m.ptr());
+  initPythonIRBindings(m.ptr());
+  tracer::initPythonTracerBindings(m.ptr());
+  top.def(
```
nit: do we want to add a comment here with the issue as well? Explaining why this was moved outside of the tracer init?

SplitInfinity replied:
I added a comment here too. This one is actually fixable because the use of `torch._C._get_tracing_state` that we need to fix is in torchvision.

albanD reviewed (approved):
Thanks for doing all of that!

Just small small comments to ease the follow up work.

Hopefully there won't be too many internal usage that needs fixing :)

Commit:
Update on "[JIT] Create torch._C._jit submodule for JIT bindings"


**Summary**
Everything inside `torch._C` is currently added to `torch` (with some
policies to exclude names beginning with `_`, etc.). As a result, some
of the JIT Python IR bindings are inadvertently exposed as `torch.XX`.
This commit adds a `torch._C._jit` submodule for all JIT bindings to
preclude JIT bindings being added to `torch`.

**Test Plan**
Continuous integration.

**Fixes**
This commit fixes #51691.

[ghstack-poisoned]

Commit:
Update on "[JIT] Create torch._C._jit submodule for JIT bindings"


**Summary**
Everything inside `torch._C` is currently added to `torch` (with some
policies to exclude names beginning with `_`, etc.). As a result, some
of the JIT Python IR bindings are inadvertently exposed as `torch.XX`.
This commit adds a `torch._C._jit` submodule for all JIT bindings to
preclude JIT bindings being added to `torch`.

**Test Plan**
Continuous integration.

**Fixes**
This commit fixes #51691.

Differential Revision: [D26628027](https://our.internmc.facebook.com/intern/diff/D26628027)

[ghstack-poisoned]

SplitInfinity commented:
@albanD I'll post another PR with the approach we discussed offline.
